Red-Black Tree/RedBlackTree.playground/Sources/RedBlackTree.swift

```diff
@@ -0,0 +1,751 @@
+//Copyright (c) 2016 Matthijs Hollemans and contributors
+//
+//Permission is hereby granted, free of charge, to any person obtaining a copy
+//of this software and associated documentation files (the "Software"), to deal
+//in the Software without restriction, including without limitation the rights
+//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//copies of the Software, and to permit persons to whom the Software is
+//furnished to do so, subject to the following conditions:
+//
+//The above copyright notice and this permission notice shall be included in
+//all copies or substantial portions of the Software.
+//
+//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//THE SOFTWARE.
+
+import Foundation
+
+private enum RBTreeColor {
+  case red
+  case black
+}
+
+private enum RotationDirection {
+  case left
+  case right
+}
+
+// MARK: - RBNode
+
+public class RBTreeNode<Key: Comparable>: Equatable {
+  public typealias RBNode = RBTreeNode<Key>
+  
+  fileprivate var color: RBTreeColor = .black
+  fileprivate var key: Key?
+  fileprivate var isNullLeaf = true
```

Comment:
Is it necessary to have a variable for this? isNullLeaf just checks to see if it's key, leftchild,rightChild, and parent is nil right? If we could remove this, that's one less variable to maintain. We could create a computed property, or func tion to check if it's a nullLeaf maybe?